src/span.js

```diff
@@ -189,81 +254,60 @@ export default class Span {
             }
         }
 
-        if (!this._imp) {
-            return;
-        }
-        this._imp.finish(finishTime);
-    }
+        this._finish(finishTime);
 
-    /**
-     * Sets a key:value pair on this Span that also propagates to future
-     * children of the associated Span.
-     *
-     * setBaggageItem() enables powerful functionality given a full-stack
-     * opentracing integration (e.g., arbitrary application data from a web
-     * client can make it, transparently, all the way into the depths of a
-     * storage system), and with it some powerful costs: use this feature with
-     * care.
-     *
-     * IMPORTANT NOTE #1: setBaggageItem() will only propagate baggage items to
-     * *future* causal descendants of the associated Span.
-     *
-     * IMPORTANT NOTE #2: Use this thoughtfully and with care. Every key and
-     * value is copied into every local *and remote* child of the associated
-     * Span, and that can add up to a lot of network and cpu overhead.
-     *
-     * @param {string} key
-     * @param {string} value
-     */
-    setBaggageItem(key, value) {
-        if (process.env.NODE_ENV === 'debug') {
-            if (arguments.length !== 2) {
-                throw new Error('Invalid number of arguments');
-            }
-        }
-        if (this._imp) {
-            this._imp.setBaggageItem(key, value);
-        }
+        // Do not return this. The Span generally should not be used after it
+        // is finished so chaining is not desired in this context.
     }
 
-    /**
-     * Returns the value for a baggage item given its key.
-     *
-     * @param  {string} key
-     *         The key for the given trace attribute.
-     * @return {string}
-     *         String value for the given key, or undefined if the key does not
-     *         correspond to a set trace attribute.
-     */
-    getBaggageItem(key) {
-        if (process.env.NODE_ENV === 'debug') {
-            if (arguments.length !== 1) {
-                throw new Error('Invalid number of arguments');
-            }
-        }
-        if (this._imp) {
-            return this._imp.getBaggageItem(key);
-        }
-        return undefined;
-    }
 
     // ---------------------------------------------------------------------- //
-    // Private and non-standard methods
+    // Methods to be implemented by derived classes
     // ---------------------------------------------------------------------- //
 
-    /**
-     * Constructs a new Span object, this method should not be called directly;
-     * Tracer.startSpan() or Tracer.join() should be used instead.
-     */
-    constructor(imp) {
-        this._imp = imp;
+    // By default returns a no-op SpanContext.
+    _context() {
+        return new SpanContext();
```

Comment:
Good point. That said, I haven't been paying attention to the details so much as I primarily wanted to make sure this general refactoring met the requirements you had in mind.